evm/chains/sharding/mainchain_handler/mainchain_handler.py

```diff
@@ -0,0 +1,78 @@
+import time
+
+import rlp
+
+from eth_utils import (
+    to_checksum_address,
+)
+
+from evm.chains.sharding.mainchain_handler.config import (
+    GASPRICE,
+    TX_GAS,
+)
+
+
+class MainchainHandler:
+
+    def __init__(self, web3_instance):
+        self.w3 = web3_instance
+        assert self.w3.isConnected()
+
+    # RPC related
+
+    def get_block_by_number(self, block_number):
+        return self.w3.eth.getBlock(block_number)
+
+    def get_block_number(self):
+        return self.w3.eth.blockNumber
+
+    def get_code(self, address):
+        address = to_checksum_address(address)
+        return self.w3.eth.getCode(address)
+
+    def get_nonce(self, address):
+        address = to_checksum_address(address)
+        return self.w3.eth.getTransactionCount(address)
+
+    def mine(self, number):
+        expected_block_number = self.get_block_number() + number
+        self.w3.miner.start(1)
+        while self.get_block_number() < expected_block_number:
+            time.sleep(0.1)
+        self.w3.miner.stop()
+
+    def unlock_account(self, account, passphrase):
+        account = to_checksum_address(account)
+        self.w3.personal.unlockAccount(account, passphrase)
```

Comment:
Unlocking accounts is generally considered *bad*, especially over RPC where *anything* can get at it.  What do you think about removing the account unlocking and instead doing one of the following.

- Issue transactions over `web3.personal.sendTransaction`
- Using the [`web3.eth.account`](http://web3py.readthedocs.io/en/latest/web3.eth.account.html) API to sign transactions locally and send them over `web3.eth.sendRawTransaction`.  This should allow for code that is **significantly** more robust and should work more reliably across different backends.

Reply:
@pipermerriam  I think you're right. It is more dangerous to use `unlockAccount` over other methods.
I'd like to change to any of the two solutions. However, it seems inevitable that we use `unlockAccount` when we do `eth.contract.transact()` using `eth.contract` if I don't understand it wrong.
In this condition can we get rid of `unlockAccount` with any solutions?
Thank you so much!

Reply:
I think that we can still get around this using the following APIs that I *think* are all avialable in web3.py today.

* Use `web3.eth.contract.buildTransaction()` to build a transaction dictionary
* Use `web3.eth.account.signTransaction()` to convert the txn dict to a raw transaction.
* Use `web3.eth.sendRawTransaction()` to send the transaction.

This approach should work unilaterally across all of the different RPC servers out there (with the assumption that the fixes for `eth-tester:py-evm` are in place.)

Reply:
Wow it's cool. I didn't even know this method `buildTransaction`. I Will try it right away. Really appreciate!